File: engine/astra/rpc_handlers/frontier.py
"""Roadmap "frontier" domains: habitability scoring/ranking, NEO hazard
assessment, asteroseismology, biosignature synthesis/fit/detection, and
technosignature search/cadence.

Split out of rpc.py (see that module's docstring for why); nothing here
changed behavior, only location.
"""

from __future__ import annotations

from typing import Any

from .common import Handler

from .. import (asteroseismology, biosignature, biosignature_fit, evidence,
                exoplanet_archive, habitability, neo_hazard, security,
                significance, store, technosignature)

def _handle_habitability_score(params: dict[str, Any]) -> dict[str, Any]:
    """Score a named confirmed planet's HZ position and ESI.

    Diagnostic evidence only -- see `habitability.py`'s module docstring
    -- never folded into `evidence.WEIGHTS`/`scoring.combine()`, matching
    every other read-only interpretation handler in this file
    (`physical.characterize`, `significance.calibrate`).
    """
    return habitability.score_archive_planet(
        str(params["planet_name"]), offline=bool(params.get("offline", False)))


def _handle_habitability_rank(params: dict[str, Any]) -> dict[str, Any]:
    records = exoplanet_archive.query_planets_bounded(
        teff_min=params.get("teff_min"), teff_max=params.get("teff_max"),
        insolation_min=params.get("insolation_min"), insolation_max=params.get("insolation_max"),
        max_rows=int(params.get("max_rows", 500)), offline=bool(params.get("offline", False)))
    ranked = habitability.rank_planets(records, limit=int(params.get("limit", 50)))
    return {"count": len(ranked), "planets": ranked}


def _handle_neo_assess(params: dict[str, Any]) -> dict[str, Any]:
    """Hazard assessment for one object's orbital elements.

    `elements` follows the same shape `moving_objects.state_vector_to_elements`
    already produces. `earth_elements` defaults to a circular 1 AU orbit when
    omitted, so a caller need not compute Earth's own osculating elements
    just to get a MOID -- diagnostic evidence only, never folded into
    `evidence.WEIGHTS`/`scoring.combine()`, per `neo_hazard.py`'s docstring.
    """
    elements = dict(params["elements"])
    earth_elements = params.get("earth_elements")
    if earth_elements is None:
        earth_elements = {"semi_major_axis_au": 1.0, "eccentricity": 0.0167,
                          "inclination_deg": 0.0, "raan_deg": 0.0,
                          "argument_of_perihelion_deg": 0.0, "mean_anomaly_deg": 0.0,
                          "epoch_mjd": elements.get("epoch_mjd", 60000.0)}
    return neo_hazard.assess(
        elements, earth_elements=earth_elements,
        apparent_v=params.get("apparent_v"), heliocentric_au=params.get("heliocentric_au"),
        geocentric_au=params.get("geocentric_au"), phase_angle_deg=params.get("phase_angle_deg"))


def _handle_neo_close_approach(params: dict[str, Any]) -> dict[str, Any]:
    return neo_hazard.close_approach(
        dict(params["elements"]), start_mjd=float(params["start_mjd"]),
        end_mjd=float(params["end_mjd"]), step_days=float(params.get("step_days", 1.0)))


def _handle_asteroseismology_measure(params: dict[str, Any]) -> dict[str, Any]:
    """Measure numax/Dnu (and, if `teff_k` given, R/M) from a stored curve.

    Reads the same `path`-addressed stored-curve convention `curves.get`
    already uses (`store.read_curve`), rather than shipping a raw
    thousands-of-points array through the JSON-lines RPC transport.
    """
    curve = store.read_curve(security.authorized_path(params["path"])).dropna().sorted_by_time()
    teff_k = params.get("teff_k")
    return asteroseismology.measure(curve.time, curve.value,
                                    teff_k=float(teff_k) if teff_k is not None else None)


def _handle_technosignature_search(params: dict[str, Any]) -> dict[str, Any]:
    """Synthesize a demo dynamic spectrum and run the de-Doppler search.

    Every number here is measured on SYNTHETIC data -- see
    `technosignature.py`'s module docstring `[GAP]`: there is no real
    Breakthrough Listen data path. Diagnostic-only, never folded into
    `evidence.WEIGHTS`/`scoring.combine()`.
    """
    synthesis = technosignature.synthesize_waterfall(
        n_time=int(params.get("n_time", 16)), n_freq=int(params.get("n_freq", 1024)),
        f0_hz=float(params.get("f0_hz", 1.4e9)),
        channel_width_hz=float(params.get("channel_width_hz", 2.7939677)),
        dt_s=float(params.get("dt_s", 18.25)),
        drift_rate_hz_s=float(params.get("drift_rate_hz_s", 0.0)),
        snr=float(params.get("snr", 0.0)),
        start_channel=params.get("start_channel"),
        seed=int(params.get("seed", 42)))
    spectrum = synthesis["spectrum"]
    result = technosignature.search(
        spectrum, max_drift_hz_s=float(params.get("max_drift_hz_s", technosignature.DEFAULT_MAX_DRIFT_HZ_S)),
        snr_threshold=float(params.get("snr_threshold", technosignature.DEFAULT_SNR_THRESHOLD)))
    result["truth"] = synthesis["truth"]
    return result


def _handle_technosignature_cadence(params: dict[str, Any]) -> dict[str, Any]:
    def _to_hits(rows: list[dict[str, Any]]) -> list[technosignature.TechnosignatureHit]:
        return [technosignature.TechnosignatureHit(
            frequency_hz=float(row["frequency_hz"]), drift_rate_hz_s=float(row["drift_rate_hz_s"]),
            snr=float(row["snr"]), freq_channel_index=int(row["freq_channel_index"]),
            drift_index=int(row["drift_index"])) for row in rows]

    on_scans = [_to_hits(scan) for scan in params["on_hit_lists"]]
    off_scans = [_to_hits(scan) for scan in params["off_hit_lists"]]
    survivors = technosignature.cadence_filter(
        on_scans, off_scans, frequency_tolerance_hz=float(params.get("frequency_tolerance_hz", 1.0)),
        drift_tolerance_hz_s=float(params.get("drift_tolerance_hz_s", 0.1)))
    return {"survivors": [hit.to_dict() for hit in survivors], "n_survivors": len(survivors)}


def _handle_biosignature_synthesize(params: dict[str, Any]) -> dict[str, Any]:
    """Generate a synthetic transmission spectrum for the panel's demo
    workflow -- pure forward-model evaluation plus Gaussian noise, the
    same "synthetic injection demonstration" shape `digital_twin.sample`
    already uses for a data type this engine has no live ingestion for.
    """
    import numpy as np

    system = biosignature.SystemParameters(
        stellar_radius_rsun=float(params["stellar_radius_rsun"]),
        planet_mass_mjup=float(params["planet_mass_mjup"]))
    atmosphere = biosignature.AtmosphereParameters(
        temperature_k=float(params.get("temperature_k", 1000.0)),
        mean_molecular_weight=float(params.get("mean_molecular_weight", 2.3)),
        reference_radius_rjup=float(params.get("reference_radius_rjup", 1.0)),
        abundances=tuple((str(m), float(a)) for m, a in params.get("abundances", [])))
    cross_sections = {str(k): float(v) for k, v in params.get("cross_sections", {}).items()}
    n_points = int(params.get("n_points", 40))
    wave_min, wave_max = float(params.get("wavelength_min_um", 1.0)), float(params.get("wavelength_max_um", 2.5))
    wavelength_um = np.linspace(wave_min, wave_max, n_points)
    depth = biosignature.transit_depth(wavelength_um, atmosphere, system, cross_sections=cross_sections)
    error_ppm = float(params.get("error_ppm", 50.0))
    rng = np.random.default_rng(int(params.get("seed", 42)))
    noisy_depth = depth + rng.normal(0.0, error_ppm * 1e-6, size=depth.shape)
    error = np.full_like(depth, error_ppm * 1e-6)
    return {"wavelength_um": wavelength_um.tolist(), "depth": noisy_depth.tolist(),
           "error": error.tolist(), "truth_depth": depth.tolist()}


def _handle_biosignature_fit(params: dict[str, Any]) -> dict[str, Any]:
    """Fit a transmission spectrum for one or more molecular bands.

    Diagnostic evidence only -- see `biosignature.py`'s module docstring
    -- fitted amplitudes are band-detection significance, NOT calibrated
    abundances; never folded into `evidence.WEIGHTS`/`scoring.combine()`.
    """
    system = biosignature.SystemParameters(
        stellar_radius_rsun=float(params["stellar_radius_rsun"]),
        planet_mass_mjup=float(params["planet_mass_mjup"]))
    molecules = tuple(str(m) for m in params["molecules"])
    cross_sections = {str(k): float(v) for k, v in params["cross_sections"].items()}
    result = biosignature_fit.fit_transmission_spectrum(
        params["wavelength_um"], params["depth"], params["error"], system,
        molecules=molecules, cross_sections=cross_sections,
        mean_molecular_weight=float(params.get("mean_molecular_weight", 2.3)),
        seed=int(params.get("seed", 42)))
    return result.to_dict()


def _handle_biosignature_detect(params: dict[str, Any]) -> dict[str, Any]:
    system = biosignature.SystemParameters(
        stellar_radius_rsun=float(params["stellar_radius_rsun"]),
        planet_mass_mjup=float(params["planet_mass_mjup"]))
    cross_sections = {str(k): float(v) for k, v in params["cross_sections"].items()}
    molecules = tuple(str(m) for m in params.get("molecules", list(cross_sections)))
    significances = {
        molecule: biosignature_fit.detection_significance(
            params["wavelength_um"], params["depth"], params["error"], system, molecule,
            cross_sections=cross_sections,
            mean_molecular_weight=float(params.get("mean_molecular_weight", 2.3)),
            seed=int(params.get("seed", 42)))
        for molecule in molecules
    }
    disequilibrium = biosignature_fit.disequilibrium_flag(significances)
    return {"significances": significances, "disequilibrium": disequilibrium}


def _handle_asteroseismology_solve(params: dict[str, Any]) -> dict[str, Any]:
    seismic = asteroseismology.SeismicParameters(
        numax_uhz=float(params["numax_uhz"]), delta_nu_uhz=float(params["delta_nu_uhz"]),
        teff_k=float(params["teff_k"]),
        numax_uhz_error=params.get("numax_uhz_error"), delta_nu_uhz_error=params.get("delta_nu_uhz_error"),
        teff_k_error=params.get("teff_k_error"))
    return asteroseismology.solve_scaling_relations(seismic).to_dict()


HANDLERS: dict[str, Handler] = {
    "habitability.score": _handle_habitability_score,
    "habitability.rank": _handle_habitability_rank,
    "neo.assess": _handle_neo_assess,
    "neo.close_approach": _handle_neo_close_approach,
    "asteroseismology.measure": _handle_asteroseismology_measure,
    "asteroseismology.solve": _handle_asteroseismology_solve,
    "biosignature.synthesize": _handle_biosignature_synthesize,
    "biosignature.fit": _handle_biosignature_fit,
    "biosignature.detect": _handle_biosignature_detect,
    "technosignature.search": _handle_technosignature_search,
    "technosignature.cadence": _handle_technosignature_cadence,
}

File: src-tauri/src/commands/frontier.rs
//! Roadmap "frontier" domains: habitability scoring/ranking, NEO hazard
//! assessment, asteroseismology, and biosignature/technosignature search.
//!
//! Split out of lib.rs (see commands/mod.rs for why); nothing here changed
//! behavior, only location.

use std::sync::Arc;

use serde_json::{json, Value};

use crate::engine::Engine;

use super::call_blocking;

#[tauri::command]
pub(crate) async fn engine_habitability_score(
    state: tauri::State<'_, Arc<Engine>>,
    planet_name: String,
    offline: Option<bool>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "habitability.score", json!({
        "planet_name": planet_name,
        "offline": offline.unwrap_or(false),
    })).await
}


#[tauri::command]
pub(crate) async fn engine_habitability_rank(
    state: tauri::State<'_, Arc<Engine>>,
    teff_min: Option<f64>,
    teff_max: Option<f64>,
    insolation_min: Option<f64>,
    insolation_max: Option<f64>,
    max_rows: Option<u32>,
    limit: Option<u32>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "habitability.rank", json!({
        "teff_min": teff_min,
        "teff_max": teff_max,
        "insolation_min": insolation_min,
        "insolation_max": insolation_max,
        "max_rows": max_rows.unwrap_or(500),
        "limit": limit.unwrap_or(50),
    })).await
}


#[tauri::command]
pub(crate) async fn engine_neo_assess(
    state: tauri::State<'_, Arc<Engine>>,
    elements: Value,
    earth_elements: Option<Value>,
    apparent_v: Option<f64>,
    heliocentric_au: Option<f64>,
    geocentric_au: Option<f64>,
    phase_angle_deg: Option<f64>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "neo.assess", json!({
        "elements": elements,
        "earth_elements": earth_elements,
        "apparent_v": apparent_v,
        "heliocentric_au": heliocentric_au,
        "geocentric_au": geocentric_au,
        "phase_angle_deg": phase_angle_deg,
    })).await
}


#[tauri::command]
pub(crate) async fn engine_neo_close_approach(
    state: tauri::State<'_, Arc<Engine>>,
    elements: Value,
    start_mjd: f64,
    end_mjd: f64,
    step_days: Option<f64>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "neo.close_approach", json!({
        "elements": elements,
        "start_mjd": start_mjd,
        "end_mjd": end_mjd,
        "step_days": step_days.unwrap_or(1.0),
    })).await
}


#[tauri::command]
pub(crate) async fn engine_asteroseismology_measure(
    state: tauri::State<'_, Arc<Engine>>,
    path: String,
    teff_k: Option<f64>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "asteroseismology.measure", json!({
        "path": path,
        "teff_k": teff_k,
    })).await
}


#[tauri::command]
pub(crate) async fn engine_asteroseismology_solve(
    state: tauri::State<'_, Arc<Engine>>,
    numax_uhz: f64,
    delta_nu_uhz: f64,
    teff_k: f64,
    numax_uhz_error: Option<f64>,
    delta_nu_uhz_error: Option<f64>,
    teff_k_error: Option<f64>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "asteroseismology.solve", json!({
        "numax_uhz": numax_uhz,
        "delta_nu_uhz": delta_nu_uhz,
        "teff_k": teff_k,
        "numax_uhz_error": numax_uhz_error,
        "delta_nu_uhz_error": delta_nu_uhz_error,
        "teff_k_error": teff_k_error,
    })).await
}


#[tauri::command]
pub(crate) async fn engine_technosignature_search(
    state: tauri::State<'_, Arc<Engine>>,
    n_time: Option<u32>,
    n_freq: Option<u32>,
    f0_hz: Option<f64>,
    channel_width_hz: Option<f64>,
    dt_s: Option<f64>,
    drift_rate_hz_s: Option<f64>,
    snr: Option<f64>,
    start_channel: Option<u32>,
    max_drift_hz_s: Option<f64>,
    snr_threshold: Option<f64>,
    seed: Option<u32>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "technosignature.search", json!({
        "n_time": n_time.unwrap_or(16),
        "n_freq": n_freq.unwrap_or(1024),
        "f0_hz": f0_hz.unwrap_or(1.4e9),
        "channel_width_hz": channel_width_hz.unwrap_or(2.7939677),
        "dt_s": dt_s.unwrap_or(18.25),
        "drift_rate_hz_s": drift_rate_hz_s.unwrap_or(0.0),
        "snr": snr.unwrap_or(0.0),
        "start_channel": start_channel,
        "max_drift_hz_s": max_drift_hz_s.unwrap_or(4.0),
        "snr_threshold": snr_threshold.unwrap_or(10.0),
        "seed": seed.unwrap_or(42),
    })).await
}


#[tauri::command]
pub(crate) async fn engine_biosignature_synthesize(
    state: tauri::State<'_, Arc<Engine>>,
    stellar_radius_rsun: f64,
    planet_mass_mjup: f64,
    temperature_k: Option<f64>,
    reference_radius_rjup: Option<f64>,
    abundances: Value,
    cross_sections: Value,
    n_points: Option<u32>,
    wavelength_min_um: Option<f64>,
    wavelength_max_um: Option<f64>,
    error_ppm: Option<f64>,
    seed: Option<u32>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "biosignature.synthesize", json!({
        "stellar_radius_rsun": stellar_radius_rsun,
        "planet_mass_mjup": planet_mass_mjup,
        "temperature_k": temperature_k.unwrap_or(1000.0),
        "reference_radius_rjup": reference_radius_rjup.unwrap_or(1.0),
        "abundances": abundances,
        "cross_sections": cross_sections,
        "n_points": n_points.unwrap_or(40),
        "wavelength_min_um": wavelength_min_um.unwrap_or(1.0),
        "wavelength_max_um": wavelength_max_um.unwrap_or(2.5),
        "error_ppm": error_ppm.unwrap_or(50.0),
        "seed": seed.unwrap_or(42),
    })).await
}


#[tauri::command]
pub(crate) async fn engine_biosignature_fit(
    state: tauri::State<'_, Arc<Engine>>,
    wavelength_um: Vec<f64>,
    depth: Vec<f64>,
    error: Vec<f64>,
    stellar_radius_rsun: f64,
    planet_mass_mjup: f64,
    molecules: Vec<String>,
    cross_sections: Value,
    seed: Option<u32>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "biosignature.fit", json!({
        "wavelength_um": wavelength_um,
        "depth": depth,
        "error": error,
        "stellar_radius_rsun": stellar_radius_rsun,
        "planet_mass_mjup": planet_mass_mjup,
        "molecules": molecules,
        "cross_sections": cross_sections,
        "seed": seed.unwrap_or(42),
    })).await
}


#[tauri::command]
pub(crate) async fn engine_biosignature_detect(
    state: tauri::State<'_, Arc<Engine>>,
    wavelength_um: Vec<f64>,
    depth: Vec<f64>,
    error: Vec<f64>,
    stellar_radius_rsun: f64,
    planet_mass_mjup: f64,
    molecules: Vec<String>,
    cross_sections: Value,
    seed: Option<u32>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "biosignature.detect", json!({
        "wavelength_um": wavelength_um,
        "depth": depth,
        "error": error,
        "stellar_radius_rsun": stellar_radius_rsun,
        "planet_mass_mjup": planet_mass_mjup,
        "molecules": molecules,
        "cross_sections": cross_sections,
        "seed": seed.unwrap_or(42),
    })).await
}


